### PR TITLE
[inductor] Avoid materializing Domino CE logits

### PR DESCRIPTION
**Pull request by ahmedtaha100**
## Summary

Domino cross entropy builds corrected logits, then computes losses and predictions from both the corrected and base logits. The compiled path currently saves the full corrected log-softmax result for backward.

This patch fixes that. AOTAutograd can recompute the cheap cat/log-softmax path when Inductor can fuse cat into a backward consumer. Normal min-cut still decides whether its replacement inputs are smaller than saving cat.

Part of pytorch/pytorch#190718

## Results

The exact BF16 reproduction on an RTX 5090 matched eager and the fused reference for outputs and trainable-input gradients. It produced no full-size or 15/16-size forward allocation, with an incremental forward peak of 132,096 bytes (0.125977 MiB).

The 15 focused tests cover Domino, equal-size MM inputs, `cat -> mm`, and a fusible forward-only cat.

## Test plan

```text
python test/inductor/test_fusion_regions.py TestFusionRegionDetection.test_can_fuse_cat_as_producer
python test/inductor/test_perf.py MinCutPartitioningTests
spin quicklint -a -- --take RUFF,PYFMT,NEWLINE test/inductor/test_fusion_regions.py test/inductor/test_perf.py torch/_functorch/partitioners.py torch/_inductor/lowering.py
```

All tests and quicklint passed.

## AI assistance disclosure

I wrote and reviewed the final patch and commit messages. Codex assisted with tracing the partitioner behavior, suggesting parts of the implementation, running local checks, and revising this PR description. I reviewed the final diff, test results, and description, understand the changes, and take responsibility for this contribution. Thank you.  

cc @jansel @eellison

### CODE REVIEW

**Comment by ahmedtaha100** [comment]
While checking the two new failed CI jobs, I found an activation memory edge case in the cat/log-softmax rematerialization path and pushed `d34312a` with a stricter guard and focused regression tests. The CPU Inductor TorchBench job failed on `doctr_det_predictor`, while the CUDA job failed on both `doctr_det_predictor` and `doctr_reco_predictor`, these were unrelated model download failures before compilation. The separate TorchTitan failure also reproduces on trunk.

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -4122,6 +4122,200 @@ def classify_nodes(
...
+            or node.meta.get("has_backward_hook", False)
+            or node.meta.get("partitioner_tag") not in (None, "is_forward")
+            or isinstance(node.meta.get("memory_budget"), float)
+            or _get_memory_budget_annotation(node) is not None
+            for node in metadata_nodes
+        ):
+            continue
+
+        for node in spine:
+            node.meta["recompute"] = CheckpointPolicy.PREFER_RECOMPUTE
```
[the remark this answers is not in the document]
Comment (ahmedtaha100) [reply torch/_functorch/partitioners.py]: d34312a addresses this by applying the optimization only when the full added cut is strictly cheaper and by rejecting unsafe mixed-dtype paths. I also added focused regressions for the ordinary two-linear and mixed-dtype cases.

**Comment by jansel** [comment]
@claude review this

**Comment by eellison** [comment]
Can you please split this into two separate prs ? One for cat recomputation one for sibling fusion.

**Comment by ahmedtaha100** [comment]
@eellison done https://github.com/pytorch/pytorch/pull/194916

**Review by eellison — COMMENTED** [review]
In Inductor, cat has a real cost when its inputs are already materialized and its output must also be materialized: we allocate the full output and copy the inputs into it. Cat can be effectively free when either its producers are memory-planned directly into the cat allocation, or it is represented as `pointwise_cat` and fused into a downstream consumer so the cat output is never materialized.

For backward rematerialization, the saved cat inputs arrive as graph inputs, so producer-side memory planning is unavailable. Cat should therefore only be considered free when it has an actual fusible backward consumer. Once that is established, the normal min-cut should decide whether the replacement inputs are smaller than saving cat.

For example, two exclusive MM inputs have the same combined size as their cat, so cat should remain saved. In Domino, the base is already saved for the sibling loss, so only the smaller correction boundary is added and recomputing cat is profitable.

Concretely, could we add an FX-level `can_fuse_cat_as_producer(cat)` helper next to `lowering.cat`, where Inductor's cat behavior is defined? It would conservatively answer whether a cat moved into the backward graph can be consumed without allocating a cat output by checking the device, the configured input-count limit, and whether its uses (following views) are fusible. It cannot make the exact lowering decision, which also depends on input IR and layouts, but colocating the two makes that distinction and the need to keep them aligned clear. The partitioner could use this helper to allow cat recomputation; non-fusible uses such as `cat -> mm` would remain materialized, and min-cut would still reject the change when saving cat is no larger than saving its boundaries. This models Inductor's cat profitability without a CE-specific matcher or forced recompute tags.

**Comment by ahmedtaha100** [comment]
@eellison Thank you for the feedback. I added `can_fuse_cat_as_producer(cat)` next to `lowering.cat`, and the partitioner now uses it to find cats with a fusible backward consumer. The normal min-cut then decides whether saving the cat or its replacement boundaries uses less memory. `cat -> mm` remains materialized, the equal-size MM case keeps the cat saved, and Domino recomputes the cat. I also removed the CE-specific matcher, config flag, and forced recompute tags.
